### PR TITLE
Add authoring support for @TabNavigator directive

### PR DESCRIPTION
**Pull request by ethan-kusters**
Bug/issue #, if applicable: rdar://97739641

## Summary

Add support for the tab navigator directive as described here: https://forums.swift.org/t/supporting-more-dynamic-content-in-swift-docc-reference-documentation/59527#tabnavigator-14

## Dependencies

- https://github.com/apple/swift-docc-render/pull/403

### CODE REVIEW

**Comment by ethan-kusters** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/Indexing/RenderIndexTests.swift [hunk 1]
```diff
@@ -657,6 +657,11 @@ final class RenderIndexTests: XCTestCase {
                         "title" : "My Article",
                         "icon" : "plus.svg",
                         "type" : "article"
+                      },
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/Indexing/RenderIndexTests.swift:660]: I think the test is now poorly named, since it tests more than CustomIcons at this point. In my mind there are two alternatives. Either we break off the tests into individual tests and instead of `XCTAssertEqual` on the full string, we assert on the presence of individual elements. alternatively we just rename this test to `testRenderIndexGenerationWithBookLikeContent` although I am not sure I like to refer to all of these as being book-like since I can see things like this new tab navigator being useful for large amounts of related screenshots in an article within conceptual documentation for an app or something.

Comment (ethan-kusters) [reply Tests/SwiftDocCTests/Indexing/RenderIndexTests.swift:660]: Ah I disagree. I think it's still only testing custom icons- if the navigator index doesn't include just regular article pages in it then something has gone _really_ wrong.

I just had to update the test to reflect the content of the DocC catalog since I added an additional page for a different test in `RenderNodeTranslatorTests` but the main thing this is testing for is still the special page that customizes an icon.

File: Tests/SwiftDocCTests/Semantics/Reference/TabNavigatorTests.swift [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    
+    func testInvalidParametersAndContent() throws {
+        let (renderBlockContent, problems, tabNavigator) = try parseDirective(TabNavigator.self) {
+            """
+            @TabNavigator(tabs: 3) {
+                @Tab("hi") {
+                    Hello there.
+                }
+            
+                @Tab("hey", bestTab: true) {
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/Semantics/Reference/TabNavigatorTests.swift:74]: Not sure this kind of test belongs in here. It's more checking the directive parser than the tab navigator directive itself. I think as you already went through so much effort to make directive arguments to content so declarative IMHO there is no point checking random keys like that.

File: Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift [hunk 3]
```diff
@@ -138,11 +138,14 @@ extension XCTestCase {
         var analyzer = SemanticAnalyzer(source: nil, context: context, bundle: bundle)
         let result = analyzer.visit(blockDirectiveContainer)
         
-        let problemIDs = analyzer.problems.map { problem -> String in
-            let line = problem.diagnostic.range?.lowerBound.line.description ?? "unknown-line"
+        let problemIDs = try analyzer.problems.map { problem -> (line: Int, id: String) in
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:141]: This doesn't feel like it's specific to this change, maybe split it up into a separate commit?

File: Tests/SwiftDocCTests/Indexing/RenderIndexTests.swift (hunk 1 above)
Comment (daniel-grumberg) [reply Tests/SwiftDocCTests/Indexing/RenderIndexTests.swift:660]: I see what you mean, I think it's fine for this review then. FTR though I think that for someone new to the code base without the context of this PR, it is less clear what the test is checking for. We should come up with a better way of testing individual render index features without having to do this all the time.

File: Tests/SwiftDocCTests/Semantics/Reference/TabNavigatorTests.swift (hunk 2 above)
Comment (ethan-kusters) [reply Tests/SwiftDocCTests/Semantics/Reference/TabNavigatorTests.swift:74]: Hm yeah that's fair. I'm asserting on the generated RenderBlockContent when an unknown directive is encountered as well below which is a little more interesting but I agree that `bestTab: true` is kind of useless at this point. I can remove it.

Comment (daniel-grumberg) [reply Tests/SwiftDocCTests/Semantics/Reference/TabNavigatorTests.swift:74]: The unknown directive case is more interesting in the sense that it maps more closely to the structure we expect from the content specifically for tab navigators.

File: Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift (hunk 3 above)
Comment (ethan-kusters) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:141]: I thought about it but it's needed for the test changes in this PR so I think it belongs here.

Some of the diagnostics being asserted on in `TabNavigatorTests` go beyond line `10` so they were being mis-sorted (`11 -` coming before `1 -`). This just resolves it by sorting on the line number specifically instead of only the string.

Comment (daniel-grumberg) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:141]: OK I see, fair enough.

File: Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift [hunk 4]
```diff
@@ -138,11 +138,14 @@ extension XCTestCase {
...
+            let line = try XCTUnwrap(problem.diagnostic.range).lowerBound.line
             
-            return "\(line): \(problem.diagnostic.severity) – \(problem.diagnostic.identifier)"
-        }.sorted()
+            return (line, "\(line): \(problem.diagnostic.severity) – \(problem.diagnostic.identifier)")
+        }
+        .sorted(by: \.id)
```
Comment (daniel-grumberg) [inline Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:146]: The double sort feels really confusing. there must be a better way of achieving this. If we just need them sorted by line number then let's just keep the second sort. If we then need ordering on other criteria let's make it explicit and do a simple, `.sorted`.

Comment (ethan-kusters) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:146]: Ah I'm open to suggestions. We need it sorted on the line number so that the tests are readable but then we need the sort to be stable between test runs for multiple diagnostics that appear on the same line.

I considered dropping into a full `sorted` but I think those are inherently much more confusing to read.

Comment (ethan-kusters) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:146]: Oh I just hadn't considered sorting the tuple itself – will that work? Agreed that's a much nicer solution if so.

Comment (ethan-kusters) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift:146]: Hm yeah a (Int, String) isn't comparable so just `sorted()` doesn't work unfortunately. I'll implement it manually and see how it is.

Comment (ethan-kusters) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift]: Okay I moved to a custom `.sorted`. Let me know what you think.

**Comment by ethan-kusters** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift (hunk 4 above)
Comment (daniel-grumberg) [reply Tests/SwiftDocCTests/XCTestCase+LoadingTestData.swift]: Interestingly enough you could write the sorted as `.sorted { $0 < $1 } because tuples somehow implement `<` without conforming to comparable apparently. I still think your version is better though

**Comment by daniel-grumberg** [comment]
LGTM

**Comment by ethan-kusters** [comment]
@swift-ci please test

**Review by daniel-grumberg — APPROVED** [review]

**Comment by ethan-kusters** [comment]
@swift-ci please test
